homeassistant/setup.py

```diff
@@ -168,14 +172,22 @@ async def _async_process_dependencies(
 
     Returns a list of dependencies which failed to set up.
     """
+    setup_futures: dict[str, asyncio.Future[bool]] = hass.data.setdefault(
+        DATA_SETUP, {}
+    )
+
     dependencies_tasks = {
-        dep: hass.loop.create_task(async_setup_component(hass, dep, config))
+        dep: setup_futures.get(dep)
+        or hass.loop.create_task(
+            async_setup_component(hass, dep, config),
+            name=f"setup {dep} as dependency of {integration.domain}",
+        )
```

Comment:
We still loose the race sometimes and end up creating the task which than waits like it did before.